kv: fix duplicate insert problem

spencerkimball opened the pull request:
If a Raft re-proposal results in an error, escalate the error to an
`AmbiguousResultError` to ensure that we properly report to the client
that despite a failure, the original request may have actually succeeded.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10703)
<!-- Reviewable:end -->

spencerkimball committed:
kv: fix Jepsen duplicate insert problem

If a Raft re-proposal results in an error, escalate the error to an
`AmbiguousResultError` to ensure that we properly report to the client
that despite a failure, the original request may have actually succeeded.

This is done only in the case of the batch containing an EndTransaction
request. Unittest added to verify operation for 1PC, single EndTxn with
both commit={true,false}, as well as non-transactional puts.

Fixes #10640

Jepsen test invocation:

```
lein do clean, run test        --tarball https://storage.googleapis.com/cockroachspencer/cockroach.tgz       --username admin  --nodes-file ~/nodes        --test sets        --nemesis majority-ring        --nemesis2 subcritical-skews        --time-limit 180  --test-count 100
```

spencerkimball commented on pkg/storage/replica.go:
```diff
@@ -1640,8 +1640,39 @@ func (r *Replica) assert5725(ba roachpb.BatchRequest) {
 // to allow commands which commute to be proposed in parallel. The naive
-// alternative (submitting requests to Raft one after another, paying massive
-// latency) is only taken for commands whose effects may overlap.
+// alternative, submitting requests to Raft one after another, paying massive
```
Note that a lot of this change is merely moving comments from `addWriteCmd` to `tryAddWriteCmd`.

spencerkimball commented on pkg/storage/replica.go:
```diff
@@ -1640,8 +1640,39 @@ func (r *Replica) assert5725(ba roachpb.BatchRequest) {
+			// If we've retried and fetched an error, we must return an
+			// ambiguous result as we can't be sure of the ultimate success
+			// of earlier attempts.
+			if count > 0 && pErr != nil {
```
This is the actual change.

spencerkimball commented:
This change ran on Jepsen for 100 iterations of the "sets" test suite. It previously wouldn't go for more than 5 without failing. Jepsen test command as follows:

```
lein do clean, run test        --tarball https://storage.googleapis.com/cockroachspencer/cockroach.tgz       --username admin  --nodes-file ~/nodes        --test sets        --nemesis majority-ring        --nemesis2 subcritical-skews        --time-limit 180  --test-count 100
```

tamird commented:
`jepsen` is not a package; the commit message and PR subject lines should start with `storage:`.

Also, can you include the jepsen invocation in the commit message?

Can you add a test that exercises this condition?

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1646 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWdS5ernpzv01T6EFmT:-KWdS5esK_uzvmVRTM6F:bbzsbvq) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1646)):*
> ```Go
> // latency, is only taken for commands whose effects may overlap.
> //
> // Internally, multiple iterations of the above process are may take place due
> ```

this sentence is mangled.

---

*[pkg/storage/replica.go, line 1660 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWdTbEZ5DH-fFlq6eoa:-KWdTbEZ5DH-fFlq6eob:bb9dlok) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1660)):*
> ```Go
> 			if count > 0 && pErr != nil {
> 				if _, ok := pErr.GetDetail().(*roachpb.AmbiguousResultError); !ok {
> 					log.ErrEventf(ctx, "received error after retries; returning ambiguous result: %v", pErr)
> ```

s/%v/%s/

---

*[pkg/storage/replica.go, line 1660 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWdTflTy4UXMkU6za1g:-KWdTflTy4UXMkU6za1h:bqo2ghn) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1660)):*
> ```Go
> 			if count > 0 && pErr != nil {
> 				if _, ok := pErr.GetDetail().(*roachpb.AmbiguousResultError); !ok {
> 					log.ErrEventf(ctx, "received error after retries; returning ambiguous result: %v", pErr)
> ```

include the number of retries since you have it?

---

*[pkg/storage/replica.go, line 1669 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWdRwNxyf1BPsVs1EmB:-KWdRwNxyf1BPsVs1EmC:br6k532) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1669)):*
> ```Go
> }
> 
> // tryAddWriteCmd implements the logic outlined in its caller addWriteCmd, who
> ```

the logic is now outlined here, not in the caller.

---

*[pkg/storage/replica.go, line 1670 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWdTMT84RVeDonnimar:-KWdTMT84RVeDonnimas:boprkqc) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1670)):*
> ```Go
> 
> // tryAddWriteCmd implements the logic outlined in its caller addWriteCmd, who
> // will call this method until the returned boolean is false, in which case
> ```

s/the returned boolean/shouldRetry/ (it's a named return)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-:-KWdRPrbRKKVOuLQK76g:b-5cngra)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1657 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWdZVifIWhe9TGT4BJA:-KWdZVifIWhe9TGT4BJB:b-mzkx0b) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1657)):*
> ```Go
> 			// If we've retried and fetched an error, we must return an
> 			// ambiguous result as we can't be sure of the ultimate success
> 			// of earlier attempts.
> ```

This sentence reads awkwardly to me. How about: "If we encounter an error on a retry we must return an ambiguous result as we can't be sure of the ultimate success of earlier attempts."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-:-KWdZp07HA1-aGeB2bD-:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1656 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWfQUO3lRxU4307140-:-KWfQUO3lRxU43071400:bx6njfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1656)):*
> ```Go
> 		if !shouldRetry {
> 			// If we've retried and fetched an error, we must return an
> 			// ambiguous result as we can't be sure of the ultimate success
> ```

Why can't we be sure? If `count > 0`, then the previous attempts returned `shouldRetry=true`. The comment below is clear that the previous attempt is guaranteed not to succeed in that case. Seems like we should either fix whatever is causing shouldRetry to be returned incorrectly or update that comment. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
There's a pending test in #10640, which will cover this case.

---

Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1646 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWdS5ernpzv01T6EFmT:-KWh94qTnzNhhsNRqAjz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1646)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this sentence is mangled.
</details>
Done.

---

*[pkg/storage/replica.go, line 1656 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWfQUO3lRxU4307140-:-KWhA4i4Row7oQ7W0SrP:bmp7oy8) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1656)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why can't we be sure? If `count > 0`, then the previous attempts returned `shouldRetry=true`. The comment below is clear that the previous attempt is guaranteed not to succeed in that case. Seems like we should either fix whatever is causing shouldRetry to be returned incorrectly or update that comment. 
</details>
I've revised to return enums to indicate the reason for proposal retry: either an illegal raft lease index (which can be retried safely) and a re-proposal (which can't be retried without returning ambiguous result on errors). Fixed the various comments.

---

*[pkg/storage/replica.go, line 1657 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWdZVifIWhe9TGT4BJA:-KWh96D8aRf6uhn0Q2AM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1657)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> This sentence reads awkwardly to me. How about: "If we encounter an error on a retry we must return an ambiguous result as we can't be sure of the ultimate success of earlier attempts."
</details>
Done.

---

*[pkg/storage/replica.go, line 1660 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWdTbEZ5DH-fFlq6eoa:-KWh98t-dysZKrCLtLQm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1660)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/%v/%s/
</details>
Done.

---

*[pkg/storage/replica.go, line 1660 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWdTflTy4UXMkU6za1g:-KWh9BehqS3fXJO6ZWm-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1660)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> include the number of retries since you have it?
</details>
Done.

---

*[pkg/storage/replica.go, line 1669 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWdRwNxyf1BPsVs1EmB:-KWh9osRUOcw3zWY4-sK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1669)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> the logic is now outlined here, not in the caller.
</details>
Done.

---

*[pkg/storage/replica.go, line 1670 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWdTMT84RVeDonnimar:-KWh9pl0EIdv6yJecSdS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1670)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/the returned boolean/shouldRetry/ (it's a named return)
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-:-KWh9rADLcekvSRK2eGf:b-v7cntq)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 1 of 3 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica.go, line 1656 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWfQUO3lRxU4307140-:-KWhjvC93ZnFff9sO0B0:bkwrjqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1656)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> I've revised to return enums to indicate the reason for proposal retry: either an illegal raft lease index (which can be retried safely) and a re-proposal (which can't be retried without returning ambiguous result on errors). Fixed the various comments.
</details>
Why are we trying so hard to propose the command if ultimately we're going to return an ambiguous result error regardless?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 175 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWhw6RsH7pMpOAvfhQc:-KWhw6RsH7pMpOAvfhQd:bfldnlt) ([raw file](https://github.com/cockroachdb/cockroach/blob/36643996ab17d8e8fc56baf75055cee4733fddb6/pkg/storage/replica.go#L175)):*
> ```Go
> 	// it should be retried. However, the original proposal may have succeeded,
> 	// so if the retry does not succeed, care must be taken to correctly inform
> 	// the caller via an AmbiguousResultError.
> ```

isn't an ambiguous result error returned regardless of the retry's success?

---

*[pkg/storage/replica.go, line 1585 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWhvhWPdKCGWEbs2iR2:-KWhvhWPdKCGWEbs2iR3:b-vkei08) ([raw file](https://github.com/cockroachdb/cockroach/blob/36643996ab17d8e8fc56baf75055cee4733fddb6/pkg/storage/replica.go#L1585)):*
> ```Go
> 
> 	endCmdsFunc := func(
> 		_ *roachpb.BatchResponse, pErr *roachpb.Error, retry proposalRetryReason,
> ```

s/retry/_/

---

*[pkg/storage/replica.go, line 1678 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWhvSPH4C71B0CHeHgd:-KWhvSPH4C71B0CHeHge:b-v9tevk) ([raw file](https://github.com/cockroachdb/cockroach/blob/36643996ab17d8e8fc56baf75055cee4733fddb6/pkg/storage/replica.go#L1678)):*
> ```Go
> 			continue
> 		}
> 		if pErr == nil {
> ```

awkward, since this is already returning `br, pErr` at the bottom.

---

*[pkg/storage/replica.go, line 1684 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWhvsL5Rmf-y_L0uaZ-:-KWhv1Mt4kmg86uSUrtq:bj69gt0) ([raw file](https://github.com/cockroachdb/cockroach/blob/36643996ab17d8e8fc56baf75055cee4733fddb6/pkg/storage/replica.go#L1684)):*
> ```Go
> 		// we must return an AmbiguousResultError as we can't be sure of
> 		// the ultimate success of earlier attempts.
> 		if _, ok := pErr.GetDetail().(*roachpb.AmbiguousResultError); !ok && ambiguousResult {
> ```

if `ambiguousResult` is `true`, this method always returns an ambiguous result error - why even bother with this type assertion?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 2 of 7 files reviewed at latest revision, 8 unresolved discussions.

---

*[pkg/storage/replica.go, line 1656 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWfQUO3lRxU4307140-:-KWiCh5hnrGddxxe3Fux:b6lcqxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/abfe2fd5b4cf1e461a84ee8281162676b3e3190a/pkg/storage/replica.go#L1656)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Why are we trying so hard to propose the command if ultimately we're going to return an ambiguous result error regardless?
</details>
If it succeeds on re-proposal, we don't have to return the ambiguous result error.

---

*[pkg/storage/replica.go, line 175 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWhw6RsH7pMpOAvfhQc:-KWiCQPBN-zKJhDUIm3R:bk1vq1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/36643996ab17d8e8fc56baf75055cee4733fddb6/pkg/storage/replica.go#L175)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> isn't an ambiguous result error returned regardless of the retry's success?
</details>
No because of our various replay protections.

---

*[pkg/storage/replica.go, line 1585 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWhvhWPdKCGWEbs2iR2:-KWiCebvrv3bqqiMfpvp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36643996ab17d8e8fc56baf75055cee4733fddb6/pkg/storage/replica.go#L1585)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/retry/_/
</details>
Done.

---

*[pkg/storage/replica.go, line 1678 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWhvSPH4C71B0CHeHgd:-KWiEQyy8qHMITP_CxSE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/36643996ab17d8e8fc56baf75055cee4733fddb6/pkg/storage/replica.go#L1678)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> awkward, since this is already returning `br, pErr` at the bottom.
</details>
Done.

---

*[pkg/storage/replica.go, line 1684 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWhvsL5Rmf-y_L0uaZ-:-KWiCvGAbd-9Tvs87vjd:b-qatits) ([raw file](https://github.com/cockroachdb/cockroach/blob/36643996ab17d8e8fc56baf75055cee4733fddb6/pkg/storage/replica.go#L1684)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> if `ambiguousResult` is `true`, this method always returns an ambiguous result error - why even bother with this type assertion?
</details>
Because I didn't want to lose the original error. (I've added a message to it now to help clarify various failures you see routinely in unittests (e.g. commands being cancelled during shutdown).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 5 of 5 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
This change LGTM, but I don't think it's fair to lay the responsibility for testing this on #10640 (for one thing, that's a *desire* for a test (or more likely several tests) to be written, not a "pending" test). Just asking for "a test which reproposes a command" doesn't mean that that test would expose this issue (the two tests we already have which repropose commands didn't). Prior to this PR it was unclear that retryProposalsLocked needed to be tested separately from other kinds of reproposals. Now that you've made two changes in this area you're much better prepared to write tests for this change and #10639 than I am. 

---

Reviewed 1 of 2 files at r1, 1 of 2 files at r2, 4 of 5 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica.go, line 1685 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWkip5AYwCAjy_SRdMA:-KWkip5AYwCAjy_SRdMB:b-otk5wi) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0457d0625920e77b2317721253c820a23d5436f/pkg/storage/replica.go#L1685)):*
> ```Go
> 				log.ErrEventf(ctx, "received error after %d retries; returning ambiguous result: %s",
> 					count, pErr)
> 				return nil, roachpb.NewError(roachpb.NewAmbiguousResultError("failed Raft re-proposal"))
> ```

It would be nice to collect metrics here: of cases where `ambiguousResult` becomes true, how often do the reproposed commands succeed and how often do we return an ambiguous result error anyway? If the odds of success on reproposal are low (or if the total occurrence of reproposals is infrequent in comparison to other sources of ambiguous results), then we could simplify things by removing this reproposal (which I fear will always remain fragile and under-tested) and just return the ambiguous result error up front.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-:-KWkgbL-xgKUCEtuRsXK:b3khi2r)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
OK, I've tightened this up by returning an `AmbiguousResultError` only in cases where the batch contains an `EndTransactionRequest`. I've also added a unittest to verify this new mechanism.

Fixes #10640

tamird commented:
Leaving a comment that says "Fixes #10640" doesn't do anything; you'll need to include this text in the commit message or PR description (preferably the former so it is visible from `git {log,blame}`.

---

Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1849 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtaYt4J6Idlvo8-lzO:-KWtaYt4J6Idlvo8-lzP:b7cj1fi) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica.go#L1849)):*
> ```Go
> 				log.Warningf(ctx, "context cancellation of command %s", ba)
> 				endCmds = nil
> 				return nil, roachpb.NewError(roachpb.NewAmbiguousResultError("context cancellation")), proposalNoRetry
> ```

`ctx.Err().Error()`?

---

*[pkg/storage/replica.go, line 1864 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtatnPQj8RhIwmdyro:-KWtatnQOvVeBrZTpe5Q:bf8c53b) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica.go#L1864)):*
> ```Go
> 			if tryAbandon() {
> 				log.Warningf(ctx, "shutdown cancellation of command %s", ba)
> 				return nil, roachpb.NewError(roachpb.NewAmbiguousResultError("shutdown cancellation")), proposalNoRetry
> ```

this error surfaces to the user, is "shutdown cancellation" appropriate?

---

*[pkg/storage/replica_test.go, line 6467 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtbQX81KsPnbI8tnPX:-KWtbQX97JtopTgMmoOI:bee0zot) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6467)):*
> ```Go
> 			checkFn: func(pErr *roachpb.Error) error {
> 				if pErr != nil {
> 					return errors.Errorf("expected success on re-proposal of puts-only txn")
> ```

should include `pErr` e.g. `errors.Wrap(pErr.GoError(), "expected success on re-proposal of puts-only txn")`

---

*[pkg/storage/replica_test.go, line 6480 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtc3zV9nhJYqtBulKo:-KWtc3zV9nhJYqtBulKp:b4nc2rk) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6480)):*
> ```Go
> 				// unittest.
> 				if _, ok := pErr.GetDetail().(*roachpb.AmbiguousResultError); !ok && pErr != nil {
> 					return errors.Errorf("expected either no error or an AmbiguousResultError but got %T: %s", pErr, pErr)
> ```

the type of pErr is always `*roachpb.Error`, so logging it with %T is not useful. perhaps better to log `pErr.GetDetail()`:
```
if detail := pErr.GetDetail(); detail != nil {
  if _, ok := detail.(*roachpb.AmbiguousResultError); !ok {
    return errors.Wrapf(detail, "unexpected error %T", detail)
  }
}
return nil
```

---

*[pkg/storage/replica_test.go, line 6489 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtcneSKyuxu_DgQAlx:-KWtcneSKyuxu_DgQAly:bj8bld6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6489)):*
> ```Go
> 			checkFn: func(pErr *roachpb.Error) error {
> 				if _, ok := pErr.GetDetail().(*roachpb.AmbiguousResultError); !ok {
> 					return errors.Errorf("expected AmbiguousResultError but got %T: %s", pErr, pErr)
> ```

same as above.

---

*[pkg/storage/replica_test.go, line 6498 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtcpukxEPHku7KYhc6:-KWtcpukxEPHku7KYhc7:b-a2xieb) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6498)):*
> ```Go
> 			checkFn: func(pErr *roachpb.Error) error {
> 				if _, ok := pErr.GetDetail().(*roachpb.TransactionStatusError); !ok {
> 					t.Errorf("expected TransactionStatusError but got %T: %s", pErr, pErr)
> ```

this one too

---

*[pkg/storage/replica_test.go, line 6509 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtdXlL9HsSISII9ukM:-KWtdXlL9HsSISII9ukN:b-wj887t) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6509)):*
> ```Go
> 	errCh := make(chan error, 1)
> 
> 	setCustomProposalFn := func(expBatch roachpb.BatchRequest) {
> ```

what's the point of this function? seems like it can be inlined into the caller.

---

*[pkg/storage/replica_test.go, line 6527 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtd-GhUqPEa3EdhWcP:-KWtd-GhUqPEa3EdhWcQ:blq4dd5) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6527)):*
> ```Go
> 
> 	for i, c := range testCases {
> 		if i != 1 {
> ```

what's this about?

---

*[pkg/storage/replica_test.go, line 6539 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtd9gkk0hKzSphZdQn:-KWtd9gkk0hKzSphZdQo:bi1ryj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6539)):*
> ```Go
> 			// index to force the sort of re-proposal we're after.
> 			origLeaseAppliedIndex := tc.repl.mu.state.LeaseAppliedIndex
> 			tc.repl.mu.state.LeaseAppliedIndex = 10
> ```

why not `+=`? This seems to be making an unnecessary assumption.

---

*[pkg/storage/replica_test.go, line 6550 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtdIzuLPudaINqjNAW:-KWtdIzuLPudaINqjNAX:b-256nli) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6550)):*
> ```Go
> 		}()
> 
> 		// Install the custom proposal function (it is unset after the first .
> ```

cut-off sentence.

---

*[pkg/storage/replica_test.go, line 6556 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtbzeohPl9obApXdGz:-KWtbzeohPl9obApXdH-:bjvl6rw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6556)):*
> ```Go
> 
> 		if err := c.checkFn(pErr); err != nil {
> 			t.Errorf("test case %d: %s", i, err)
> ```

consider giving each test a name and using sub-tests instead of integer indices.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-:-KWt_vzyfPelW_6MDoyI:b-g7q1we)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 16 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica.go, line 1685 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWkip5AYwCAjy_SRdMA:-KWu8FtuhFSCriuadQ0G:br6s2ep) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0457d0625920e77b2317721253c820a23d5436f/pkg/storage/replica.go#L1685)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> It would be nice to collect metrics here: of cases where `ambiguousResult` becomes true, how often do the reproposed commands succeed and how often do we return an ambiguous result error anyway? If the odds of success on reproposal are low (or if the total occurrence of reproposals is infrequent in comparison to other sources of ambiguous results), then we could simplify things by removing this reproposal (which I fear will always remain fragile and under-tested) and just return the ambiguous result error up front.
</details>
I'm adding a TODO here. This change is already metastasizing.

---

*[pkg/storage/replica.go, line 1849 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtaYt4J6Idlvo8-lzO:-KWu9c_r54F-Fo8mUcsp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica.go#L1849)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> `ctx.Err().Error()`?
</details>
Done.

---

*[pkg/storage/replica.go, line 1864 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtatnPQj8RhIwmdyro:-KWu9dSfz8rXYFTUOqC3:b-2r60kg) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica.go#L1864)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this error surfaces to the user, is "shutdown cancellation" appropriate?
</details>
I think so. Client SQL commands fail to give you a definitive answer about your txn commit...makes sense if they tell you that it was because of the database server being shutdown. I changed it to "server shutdown" to maybe make it more palatable.

---

*[pkg/storage/replica_test.go, line 6467 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtbQX81KsPnbI8tnPX:-KWuASKPZ-LpfLbc9UvH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6467)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> should include `pErr` e.g. `errors.Wrap(pErr.GoError(), "expected success on re-proposal of puts-only txn")`
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6480 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtc3zV9nhJYqtBulKo:-KWuAowKm5hpNs2TBOrf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6480)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> the type of pErr is always `*roachpb.Error`, so logging it with %T is not useful. perhaps better to log `pErr.GetDetail()`:
> ```
> if detail := pErr.GetDetail(); detail != nil {
>   if _, ok := detail.(*roachpb.AmbiguousResultError); !ok {
>     return errors.Wrapf(detail, "unexpected error %T", detail)
>   }
> }
> return nil
> ```
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6489 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtcneSKyuxu_DgQAlx:-KWuApDViTujhC0wVFqm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6489)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> same as above.
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6498 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtcpukxEPHku7KYhc6:-KWuApe6P3Hl5pDrM8fC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6498)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this one too
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6509 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtdXlL9HsSISII9ukM:-KWuAxhqD9WvEyEiU_jR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6509)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's the point of this function? seems like it can be inlined into the caller.
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6527 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtd-GhUqPEa3EdhWcP:-KWuAxzlUZLp6eTU9vFZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6527)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's this about?
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6539 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtd9gkk0hKzSphZdQn:-KWuB6gNk4ojOUPFFMAJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6539)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> why not `+=`? This seems to be making an unnecessary assumption.
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6550 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtdIzuLPudaINqjNAW:-KWuB7TEoWi_z8l_X7Fl:b-1y7izh) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6550)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> cut-off sentence.
</details>
Removed with inlining.

---

*[pkg/storage/replica_test.go, line 6556 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtbzeohPl9obApXdGz:-KWuB9ex-f3mzDyX_TaT:b-8e13tf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6556)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> consider giving each test a name and using sub-tests instead of integer indices.
</details>
Nah, this is simpler and it makes so little difference when something goes wrong.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_test.go, line 6480 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtc3zV9nhJYqtBulKo:-KWuHVfyPunDPp77EREm:b5woj58) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6480)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Done.
</details>
the %T in my suggested code was intentional - the text representation of `detail` will already be inserted by the Wrapf() call.

---

*[pkg/storage/replica_test.go, line 6489 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtcneSKyuxu_DgQAlx:-KWuHWogJH4hdjPBPJNi:b5woj58) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6489)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Done.
</details>
the %T in my suggested code was intentional - the text representation of `detail` will already be inserted by the Wrapf() call.

---

*[pkg/storage/replica_test.go, line 6498 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtcpukxEPHku7KYhc6:-KWuHXayJhw9fEq-Hd86:b5woj58) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6498)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Done.
</details>
the %T in my suggested code was intentional - the text representation of `detail` will already be inserted by the Wrapf() call.

---

*[pkg/storage/replica_test.go, line 6527 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtd-GhUqPEa3EdhWcP:-KWuHZzjt2ctvIoRW6--:b5maicy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6527)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Done.
</details>
What's done?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_test.go, line 6480 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtc3zV9nhJYqtBulKo:-KWuIBMH0De2_9NQqxwp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6480)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> the %T in my suggested code was intentional - the text representation of `detail` will already be inserted by the Wrapf() call.
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6489 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtcneSKyuxu_DgQAlx:-KWuIBqMbl41dIYE3b66:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6489)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> the %T in my suggested code was intentional - the text representation of `detail` will already be inserted by the Wrapf() call.
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6498 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtcpukxEPHku7KYhc6:-KWuICEiFgaNl0YubJZQ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6498)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> the %T in my suggested code was intentional - the text representation of `detail` will already be inserted by the Wrapf() call.
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6527 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtd-GhUqPEa3EdhWcP:-KWuIDtfzZsAVOfIq9a2:b-cu94qz) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6527)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> What's done?
</details>
Oops, that's vestigial. Removed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but please wait for @bdarnell.

---

Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-:-KWuKWyd5SACgmgysZU7:bylj9xa)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, but I think #10640 should stay open until we have a test for the timestamp cache change (#10639) as well.

---

Reviewed 3 of 5 files at r4, 1 of 2 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_test.go, line 6556 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtbzeohPl9obApXdGz:-KWutgyopAdCCw9qjfgM:b-hwnamf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6556)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Nah, this is simpler and it makes so little difference when something goes wrong.
</details>
FWIW, the value of subtests is not so much the use of names vs numbers, but that they can tell you which test was running even when the failure is a panic or timeout. (and in tests with many calls to Errorf, you don't have to add the test index to every one)

---

*[pkg/storage/replica_test.go, line 6468 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWupUO5kDGAYyjHo0xQ:-KWupUO5kDGAYyjHo0xR:b2r061k) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f903d2bd92550f73b3dc1128b92b244b38d6bbf/pkg/storage/replica_test.go#L6468)):*
> ```Go
> 				if pErr != nil {
> 					return errors.Wrap(pErr.GoError(),
> 						"expected success on re-proposal of puts-only txn")
> ```

s/txn/batch/

Comment that this is the current behavior because we do not provide replay protection for non-transactional commands, although we might want to change that in the future. 

---

*[pkg/storage/replica_test.go, line 6480 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWusiMGBDgXZQBELxLW:-KWusiMGBDgXZQBELxLX:b-8h20z0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f903d2bd92550f73b3dc1128b92b244b38d6bbf/pkg/storage/replica_test.go#L6480)):*
> ```Go
> 				// or the original command finishes first. Both happen in this
> 				// unittest.
> 				if detail := pErr.GetDetail(); detail != nil {
> ```

The success case should be `pErr == nil`, not `pErr.GetDetail() == nil`. The protobuf code is nil-safe here but it's still weird to not check for nil until after calling GetDetail. 

---

*[pkg/storage/replica_test.go, line 6491 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWupq2stCHiN4fi7TBV:-KWupq2stCHiN4fi7TBW:b-ukn47b) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f903d2bd92550f73b3dc1128b92b244b38d6bbf/pkg/storage/replica_test.go#L6491)):*
> ```Go
> 			ba: baCommitTxn,
> 			checkFn: func(pErr *roachpb.Error) error {
> 				if detail := pErr.GetDetail(); detail != nil {
> ```

It shouldn't be possible to get `nil` here - you're trying to commit a transaction that doesn't exist, so you should get the error every time.

---

*[pkg/storage/replica_test.go, line 6502 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWut5ATBpjfhriJyfRe:-KWut5ATBpjfhriJyfRf:bjp9ol3) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f903d2bd92550f73b3dc1128b92b244b38d6bbf/pkg/storage/replica_test.go#L6502)):*
> ```Go
> 			ba: baAbortTxn,
> 			checkFn: func(pErr *roachpb.Error) error {
> 				if detail := pErr.GetDetail(); detail != nil {
> ```

Ditto. 

---

*[pkg/storage/replica_test.go, line 6542 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWuuqLDdPDby3f4FtC1:-KWuuqLDdPDby3f4FtC2:b-9j1q5w) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f903d2bd92550f73b3dc1128b92b244b38d6bbf/pkg/storage/replica_test.go#L6542)):*
> ```Go
> 			proposed <- struct{}{} // signal that the proposal has arrived
> 			// Submit the proposal with a delay, waiting for proceed channel.
> 			go func() {
> ```

Why are there two goroutines? Couldn't you move the body of the previous goroutine into this one? 

There's a slight race here - if submitProposalFn is called twice before the previous goroutine runs, the second run will spawn a second goroutine which will be stuck on the `proceed` channel (and if there's a third call, I think the whole thing deadlocks on the `proposed` channel). I think the assignment `submitProposalFn = nil` needs to happen synchronously in the proposal function itself instead of the goroutines. 

---

*[pkg/storage/replica_test.go, line 6562 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWuuPBotf0OyhVr7BZJ:-KWuuPBotf0OyhVr7BZK:b-uisyup) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f903d2bd92550f73b3dc1128b92b244b38d6bbf/pkg/storage/replica_test.go#L6562)):*
> ```Go
> 				t.Fatal(err)
> 			}
> 		default:
> ```

The submitProposalFn goroutine always writes a single value to the channel, so you should be able to read it here without a `default`. If you hit the default case, that means the goroutine is either still running and will write its error soon or it's gotten stuck (which should be handled as an error). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-:-KWuz0QBMq05qPcZumM7:b-pg0l1h)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
This tests the timestamp cache change.  The second test case in the new unittest verifies that the timestamp cache isn't updated on the first retry. Otherwise, we'd receive a `TransactionReplayError` instead of `AmbiguousResultError`. This is because it's composed of an entire txn including begin & end. The end txn updates timestamp cache and the subsequent begin checks it. Before the fix in #10639, this would result in a spurious `TransactionReplayError`.

---

Review status: 7 of 8 files reviewed at latest revision, 12 unresolved discussions.

---

*[pkg/storage/replica_test.go, line 6556 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWtbzeohPl9obApXdGz:-KX19lzkfPIThpIQW2u5:bq0qw21) ([raw file](https://github.com/cockroachdb/cockroach/blob/d258015bec25466c72ac8c48b8cb7402db03f0dd/pkg/storage/replica_test.go#L6556)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> FWIW, the value of subtests is not so much the use of names vs numbers, but that they can tell you which test was running even when the failure is a panic or timeout. (and in tests with many calls to Errorf, you don't have to add the test index to every one)
</details>
Ah, didn't realize this was a new feature of 1.7.

---

*[pkg/storage/replica_test.go, line 6468 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWupUO5kDGAYyjHo0xQ:-KX14ZkiMyRMnmEIB7Nf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f903d2bd92550f73b3dc1128b92b244b38d6bbf/pkg/storage/replica_test.go#L6468)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/txn/batch/
> 
> Comment that this is the current behavior because we do not provide replay protection for non-transactional commands, although we might want to change that in the future. 
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6480 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWusiMGBDgXZQBELxLW:-KX14z0bfPxqvT5CdYpE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f903d2bd92550f73b3dc1128b92b244b38d6bbf/pkg/storage/replica_test.go#L6480)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The success case should be `pErr == nil`, not `pErr.GetDetail() == nil`. The protobuf code is nil-safe here but it's still weird to not check for nil until after calling GetDetail. 
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6491 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWupq2stCHiN4fi7TBV:-KX15ESHdofoK0K2dV3Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f903d2bd92550f73b3dc1128b92b244b38d6bbf/pkg/storage/replica_test.go#L6491)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> It shouldn't be possible to get `nil` here - you're trying to commit a transaction that doesn't exist, so you should get the error every time.
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6502 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWut5ATBpjfhriJyfRe:-KX15EAIea_XHTXTRVEu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f903d2bd92550f73b3dc1128b92b244b38d6bbf/pkg/storage/replica_test.go#L6502)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Ditto. 
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6542 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWuuqLDdPDby3f4FtC1:-KX18_QHStSmGEH_dWhK:bwcr3j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f903d2bd92550f73b3dc1128b92b244b38d6bbf/pkg/storage/replica_test.go#L6542)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Why are there two goroutines? Couldn't you move the body of the previous goroutine into this one? 
> 
> There's a slight race here - if submitProposalFn is called twice before the previous goroutine runs, the second run will spawn a second goroutine which will be stuck on the `proceed` channel (and if there's a third call, I think the whole thing deadlocks on the `proposed` channel). I think the assignment `submitProposalFn = nil` needs to happen synchronously in the proposal function itself instead of the goroutines. 
</details>
Good points, Fixed.

---

*[pkg/storage/replica_test.go, line 6562 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KWuuPBotf0OyhVr7BZJ:-KX15Rq--goSLXIxu7hY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3f903d2bd92550f73b3dc1128b92b244b38d6bbf/pkg/storage/replica_test.go#L6562)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The submitProposalFn goroutine always writes a single value to the channel, so you should be able to read it here without a `default`. If you hit the default case, that means the goroutine is either still running and will write its error soon or it's gotten stuck (which should be handled as an error). 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-:-KX13BgT1BtCuC3gIKnE:b-8d9lkh)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 15 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_test.go, line 6490 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KX2FBTPjItcQxfwS3E_:-KX2FBTPjItcQxfwS3Ea:b-n4z2wx) ([raw file](https://github.com/cockroachdb/cockroach/blob/728905ad405fcbe947fab34902e98dd90b22438d/pkg/storage/replica_test.go#L6490)):*
> ```Go
> 				}
> 				if err, ok := pErr.GetDetail().(*roachpb.AmbiguousResultError); !ok {
> 					return errors.Wrapf(err, "unexpected error %T", err)
> ```

in the case of the failed type assertion, `err` will be nil, so this is not useful logging

https://play.golang.org/p/EvWD4fWAXW

---

*[pkg/storage/replica_test.go, line 6503 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KX2FbsJfNQ5j13UWmcM:-KX2FbsJfNQ5j13UWmcN:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/728905ad405fcbe947fab34902e98dd90b22438d/pkg/storage/replica_test.go#L6503)):*
> ```Go
> 				}
> 				if err, ok := pErr.GetDetail().(*roachpb.AmbiguousResultError); !ok {
> 					return errors.Wrapf(err, "unexpected error %T", err)
> ```

ditto

---

*[pkg/storage/replica_test.go, line 6516 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KX2FctSf1wIgl4sPh98:-KX2FctSf1wIgl4sPh99:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/728905ad405fcbe947fab34902e98dd90b22438d/pkg/storage/replica_test.go#L6516)):*
> ```Go
> 				}
> 				if err, ok := pErr.GetDetail().(*roachpb.TransactionStatusError); !ok {
> 					return errors.Wrapf(err, "unexpected error %T", err)
> ```

ditto

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 7 of 8 files reviewed at latest revision, 15 unresolved discussions.

---

*[pkg/storage/replica_test.go, line 6490 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KX2FBTPjItcQxfwS3E_:-KX2lvIwTOA8x2FmXxNp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/728905ad405fcbe947fab34902e98dd90b22438d/pkg/storage/replica_test.go#L6490)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> in the case of the failed type assertion, `err` will be nil, so this is not useful logging
> 
> https://play.golang.org/p/EvWD4fWAXW
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6503 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KX2FbsJfNQ5j13UWmcM:-KX2lwJxrPB2xFV__4fG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/728905ad405fcbe947fab34902e98dd90b22438d/pkg/storage/replica_test.go#L6503)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
Done.

---

*[pkg/storage/replica_test.go, line 6516 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KX2FctSf1wIgl4sPh98:-KX2lwvb4m6Ga8JFLU0N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/728905ad405fcbe947fab34902e98dd90b22438d/pkg/storage/replica_test.go#L6516)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r8.
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks pending.

---

*[pkg/storage/replica_test.go, line 6503 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KX2FbsJfNQ5j13UWmcM:-KX2od5UF0w7bpiGrPxL:bfw9e3y) ([raw file](https://github.com/cockroachdb/cockroach/blob/728905ad405fcbe947fab34902e98dd90b22438d/pkg/storage/replica_test.go#L6503)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Done.
</details>
Looks like "endtxn/commit=true" was removed entirely?

---

*[pkg/storage/replica_test.go, line 6516 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KX2FctSf1wIgl4sPh98:-KX2ognrfRDE_fu-_AH5:b2cfoe1) ([raw file](https://github.com/cockroachdb/cockroach/blob/728905ad405fcbe947fab34902e98dd90b22438d/pkg/storage/replica_test.go#L6516)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Done.
</details>
looks like "endtxn/commit=false" was removed entirely?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_test.go, line 6503 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KX2FbsJfNQ5j13UWmcM:-KX30XjwtqWot0Nu6rx5:b-l61zr6) ([raw file](https://github.com/cockroachdb/cockroach/blob/728905ad405fcbe947fab34902e98dd90b22438d/pkg/storage/replica_test.go#L6503)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Looks like "endtxn/commit=true" was removed entirely?
</details>
Yes. I realized they were unnecessary and don't work properly for proposer evaluated KV.

---

*[pkg/storage/replica_test.go, line 6516 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-KX2FctSf1wIgl4sPh98:-KX3119CVVo69pHr-AwD:b2nign6) ([raw file](https://github.com/cockroachdb/cockroach/blob/728905ad405fcbe947fab34902e98dd90b22438d/pkg/storage/replica_test.go#L6516)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> looks like "endtxn/commit=false" was removed entirely?
</details>
yes

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM. That comment about the timestamp cache and TransactionReplayError vs AmbiguousResultError should probably be added to the test.

---

Review status: 6 of 8 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-:-KX4trgrz7GVNG2t7k3y:b3zzp3w)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Done

---

Review status: 5 of 8 files reviewed at latest revision, 14 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703#-:-KX6BWxcQo4kA_h-YWB_:bb6un1j)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r9.
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10703)*
<!-- Sent from Reviewable.io -->
